
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../exercises/">
      
      
        <link rel="next" href="../links/">
      
      
      <link rel="icon" href="../assets/images/TDS_logo.svg">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.27">
    
    
      
        <title>Bonus code - Introduction to R for Life Sciences</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.6543a935.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#statistics-with-r" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Introduction to R for Life Sciences" class="md-header__button md-logo" aria-label="Introduction to R for Life Sciences" data-md-component="logo">
      
  <img src="../assets/images/TDS_logo.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Introduction to R for Life Sciences
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Bonus code
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/taniawyss/intro-to-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    taniawyss/intro-to-R
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Introduction to R for Life Sciences" class="md-nav__button md-logo" aria-label="Introduction to R for Life Sciences" data-md-component="logo">
      
  <img src="../assets/images/TDS_logo.svg" alt="logo">

    </a>
    Introduction to R for Life Sciences
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/taniawyss/intro-to-R" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    taniawyss/intro-to-R
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../precourse/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Precourse preparations
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../course_schedule/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Course schedule
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../materials/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Materials
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../exercises/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Exercises
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Bonus code
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Bonus code
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#statistics-with-r" class="md-nav__link">
    <span class="md-ellipsis">
      Statistics with R
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rna-sequencing-example-code" class="md-nav__link">
    <span class="md-ellipsis">
      RNA sequencing - example code
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RNA sequencing - example code">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#small-information-on-a-typical-rnaseq-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Small information on a typical RNAseq workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-for-differential-expression-analysis-with-deseq2" class="md-nav__link">
    <span class="md-ellipsis">
      Example for differential expression analysis with DESeq2
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#r-markdown" class="md-nav__link">
    <span class="md-ellipsis">
      R markdown
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../links/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Useful links
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#statistics-with-r" class="md-nav__link">
    <span class="md-ellipsis">
      Statistics with R
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rna-sequencing-example-code" class="md-nav__link">
    <span class="md-ellipsis">
      RNA sequencing - example code
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RNA sequencing - example code">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#small-information-on-a-typical-rnaseq-workflow" class="md-nav__link">
    <span class="md-ellipsis">
      Small information on a typical RNAseq workflow
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-for-differential-expression-analysis-with-deseq2" class="md-nav__link">
    <span class="md-ellipsis">
      Example for differential expression analysis with DESeq2
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#r-markdown" class="md-nav__link">
    <span class="md-ellipsis">
      R markdown
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Bonus code</h1>

<p>The following code shows more details about additional features of R, such as generating reports or running statistics. They might be useful for you.</p>
<h2 id="statistics-with-r">Statistics with R</h2>
<p>The SIB course <a href="https://github.com/sib-swiss/first-steps-with-R-training/tree/master">First Steps with R in life sciences</a> provides additional material to perform statistics with R.</p>
<p>The slides introducing statistics with R can be found <a href="https://github.com/sib-swiss/first-steps-with-R-training/blob/master/slides/First-steps-with-R_day2_afternoon.pdf">here</a> on the <a href="https://github.com/sib-swiss/first-steps-with-R-training">course material github page</a>. </p>
<p>The R code to run Wilcoxon or T tests can be found <a href="https://github.com/sib-swiss/first-steps-with-R-training/blob/master/solutions/R_practice9_solution.R">here (ex.9)</a> using source data that can be found <a href="https://github.com/sib-swiss/first-steps-with-R-training/tree/master/course_datasets">here</a>. The R code to run a linear model can be found <a href="https://github.com/sib-swiss/first-steps-with-R-training/blob/master/solutions/R_practice10_solution.R">here (ex.10)</a>. Feel free to try it out!</p>
<h2 id="rna-sequencing-example-code">RNA sequencing - example code</h2>
<h3 id="small-information-on-a-typical-rnaseq-workflow">Small information on a typical RNAseq workflow</h3>
<p><a class="md-button" href="../assets/pdf/brief_RNAseq_workflow.pdf">Download pdf</a></p>
<h3 id="example-for-differential-expression-analysis-with-deseq2">Example for differential expression analysis with DESeq2</h3>
<p>Once sequencing reads have been aligned to the reference genome, and number of reads have been summarized per gene per sample, we can use R for downstream differential gene expression analysis. Several R packages exist for DE analysis, such as <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>, <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a> or <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>. The short code below shows the initial steps of DE analysis with DESeq2, starting from a text file that contains numbers of sequencing reads per gene per sample, <em>i.e.</em> a matrix of raw counts. The samples included in the example matrix come from natural killer and T helper cells sorted from the peripheral blood of healthy donors.</p>
<p><a class="md-button" href="../assets/data/htseq_counts_NK_Th.csv">Download data</a>
<a class="md-button" href="../assets/data/DGE_DESeq2_example.R">Download script</a></p>
<div class="highlight"><pre><span></span><code><span class="c1"># DGE example with DESeq2:</span>
<span class="c1"># R version 4.1.2 (2021-11-01)</span>


<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;DESeq2&quot;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span><span class="w"> </span><span class="c1"># v 1.34.0</span>

<span class="c1"># setwd(&quot;path/to/downloadedData&quot;)</span>

<span class="n">counts_NK_Th</span><span class="o">&lt;-</span><span class="nf">read.csv</span><span class="p">(</span><span class="s">&quot;htseq_counts_NK_Th.csv&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="bp">T</span><span class="p">)</span>

<span class="n">counts_NK_Th</span><span class="o">&lt;-</span><span class="n">counts_NK_Th</span><span class="p">[</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="nf">rowSums</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">)</span><span class="o">==</span><span class="m">0</span><span class="p">)),]</span>
<span class="nf">dim</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">)</span>
<span class="c1"># [1] 38573    15</span>

<span class="c1"># build a sample metadata table:</span>
<span class="n">coldata</span><span class="o">&lt;-</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">cbind</span><span class="p">(</span><span class="n">cell_type</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;NK&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">),</span>
<span class="w">                           </span><span class="nf">rep</span><span class="p">(</span><span class="s">&quot;Th&quot;</span><span class="p">,</span><span class="w"> </span><span class="m">9</span><span class="p">)),</span>
<span class="w">               </span><span class="n">donor</span><span class="o">=</span><span class="nf">sapply</span><span class="p">(</span><span class="nf">strsplit</span><span class="p">(</span><span class="nf">colnames</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;_&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&#39;[&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">),</span>
<span class="w">               </span><span class="n">sample_id</span><span class="o">=</span><span class="nf">colnames</span><span class="p">(</span><span class="n">counts_NK_Th</span><span class="p">)))</span>
<span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="o">&lt;-</span><span class="nf">as.factor</span><span class="p">(</span><span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="p">)</span>
<span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="o">&lt;-</span><span class="nf">factor</span><span class="p">(</span><span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="p">,</span>
<span class="w">                          </span><span class="n">levels</span><span class="o">=</span><span class="nf">levels</span><span class="p">(</span><span class="n">coldata</span><span class="o">$</span><span class="n">cell_type</span><span class="p">)[</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">)])</span>

<span class="nf">head</span><span class="p">(</span><span class="n">coldata</span><span class="p">)</span>
<span class="c1">#   cell_type donor         sample_id</span>
<span class="c1"># 1        NK   S15    S15_NK_CD56dim</span>
<span class="c1"># 2        NK   S15 S15_NK_CD56bright</span>
<span class="c1"># 3        NK   S16    S16_NK_CD56dim</span>
<span class="c1"># 4        NK   S16 S16_NK_CD56bright</span>
<span class="c1"># 5        NK   S17    S17_NK_CD56dim</span>
<span class="c1"># 6        NK   S17 S17_NK_CD56bright</span>

<span class="c1"># Create DESeq object:</span>
<span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">counts_NK_Th</span><span class="p">,</span>
<span class="w">                              </span><span class="n">colData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">coldata</span><span class="p">,</span>
<span class="w">                              </span><span class="n">design</span><span class="o">=</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">donor</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">cell_type</span><span class="p">)</span><span class="w"> </span><span class="c1"># Difference between cell types, accounting for the sample pairing</span>
<span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">DESeq</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span>
<span class="nf">resultsNames</span><span class="p">(</span><span class="n">dds</span><span class="p">)</span><span class="w"> </span><span class="c1"># lists the coefficients</span>
<span class="c1"># [1] &quot;Intercept&quot;          &quot;donor_S16_vs_S15&quot;   &quot;donor_S17_vs_S15&quot;   &quot;cell_type_NK_vs_Th&quot;</span>

<span class="n">deseq2_NK_vs_Th</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">results</span><span class="p">(</span><span class="n">dds</span><span class="p">,</span><span class="w">  </span>
<span class="w">                                         </span><span class="n">alpha</span><span class="o">=</span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span>
<span class="w">                                         </span><span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cell_type&quot;</span><span class="p">,</span><span class="s">&quot;NK&quot;</span><span class="p">,</span><span class="s">&quot;Th&quot;</span><span class="p">),</span>
<span class="w">                                 </span><span class="n">cooksCutoff</span><span class="o">=</span><span class="bp">F</span><span class="p">))</span><span class="w"> </span><span class="c1"># use cooksCutoff=F only if some genes of interest do not have a calculated p-value</span>
<span class="w">                                                </span><span class="c1"># author recomendation is to use default cooksCutoff=T    </span>

<span class="nf">head</span><span class="p">(</span><span class="n">deseq2_NK_vs_Th</span><span class="p">)</span>
<span class="c1">#            baseMean log2FoldChange     lfcSE        stat       pvalue         padj</span>
<span class="c1"># TSPAN6     40.851111     -6.9583034 1.1044621 -6.30017417 2.973114e-10 8.742152e-09</span>
<span class="c1"># TNMD        0.104281     -0.1228379 3.1336380 -0.03919977 9.687311e-01           NA</span>
<span class="c1"># DPM1     2566.964652     -0.1466129 0.2211112 -0.66307338 5.072836e-01 7.133950e-01</span>
<span class="c1"># SCYL3     571.791633      0.5728065 0.4039864  1.41788547 1.562242e-01 3.515869e-01</span>
<span class="c1"># C1orf112  201.504414      0.8758449 0.5938469  1.47486651 1.402484e-01 3.263445e-01</span>
<span class="c1"># FGR      8793.900467      8.5188295 1.2025099  7.08420757 1.398422e-12 5.868549e-11</span>

<span class="n">deseq2_NK_vs_Th</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;CPS1&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">deseq2_NK_vs_Th</span><span class="p">)),]</span>
<span class="c1">#            baseMean log2FoldChange    lfcSE        stat     pvalue      padj</span>
<span class="c1"># CPS1     2.34916186    -3.56324252 1.855768 -1.92009033 0.05484649 0.1702518</span>
<span class="c1"># CPS1.IT1 0.09375824    -0.08381473 3.134376 -0.02674048 0.97866673        NA</span>
<span class="n">deseq2_NK_vs_Th</span><span class="p">[</span><span class="nf">grep</span><span class="p">(</span><span class="s">&quot;GZMB&quot;</span><span class="p">,</span><span class="w"> </span><span class="nf">rownames</span><span class="p">(</span><span class="n">deseq2_NK_vs_Th</span><span class="p">)),]</span>
<span class="c1">#      baseMean log2FoldChange     lfcSE     stat       pvalue        padj</span>
<span class="c1"># GZMB 27758.47       9.075387 0.8897603 10.19981 1.986563e-24 2.65665e-22</span>
</code></pre></div>
<h2 id="r-markdown">R markdown</h2>
<p>If a data analysis project involves many steps and generation of various plots, one of the easy and very practical ways to bundle and organize all steps of analysis together is to use R markdown files to generate PDF or html reports. These reports both display the R code used as well as the output generated, such as graphics, tables, statistical test results, &hellip;</p>
<p>The difference between an R script and an R markdown file, is that the code is organized within chunks in the R markdown file. In between the chunks, the user can write text that contains information about the analysis.</p>
<p>To create an R markdown file, go to File &gt; New File &gt; R markdown. Add a name.
This will create a new file that already has some example content. As you can see, the R code is organized in chunks highlighed in grey, with details written as free text in between the chunks. 
We can see that the pound sign (#) is used outside of the R code chunks. In this case, the # symbol does not correspond to a comment, but will indicate header levels for the titles and subtitles within your final document obtained after report generation.</p>
<p>Once the Rmd is ready, the report can be generated by hitting the &ldquo;Knit&rdquo; button at the top of the window. </p>
<figure>
  <img src="../assets/images/example_Rmd.png" width="800"/>
</figure>

<p>The example Rmd generates the following html report (saved in the same folder as the Rmd file by default), that shows both the code and the resulting output:</p>
<figure>
  <img src="../assets/images/example_html.png" width="800"/>
</figure>

<p>You can find a short video that introduces some of the principles of R markdown on <a href="https://www.youtube.com/watch?v=2YZSDGGoQzQ">Youtube</a>, from the beginning up to minute 23:30. Starting at minute 23:30, this video also introduces ggplot2.</p>
<p>If you would like to practice creating your own R markdown, modify the one that is generated with the example content when you select File &gt; New File &gt; R Markdown.</p>
<p>1) Create a new Rmd file with the following options at the top (in the top YAML instructions within the 2 dash sequences &ldquo;- - -&ldquo;)</p>
<ul>
<li>Title: «Let’s practice»</li>
<li>Author: your name</li>
<li>Select the «use current date when rendering object» option</li>
<li>Default output format: HTML</li>
</ul>
<p>2) We can repeat the exercises of this course, but this time by creating a report:
Within an R code chunk, insert the R code that you would like to execute. You can include code to create plots for example.
If you have more than 1 plot to include, create a new R code chunk for each, and add a header (preceded by the # symbols <strong>outside of the code chunks</strong>) before each plot with some suggestive plot title.</p>
<p>3) Save the Rmd file and produce the html document by «knitting» it.</p>
<!--[Download solution Rmd file](../assets/pdf/R_practice_Rmd_solution.Rmd){: .md-button }-->

<p>For tweaking your reports, such as chosing different output formats, or hiding or showing the code within the report, we recommend that you consult the R markdown documentation provided in this <a href="https://bookdown.org/yihui/rmarkdown/">Definite guide eBook</a>.</p>
<p>Another useful resource is <a href="https://rmarkdown.rstudio.com/lesson-1.html">RStudio&rsquo;s R Markdown tutorial</a>. </p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
    
  </body>
</html>