## Overview

Experiments designed to quantify gene expression often yield hundreds of genes that show statistically significant differences between two classes (two biological states, two phenotype states, two experimental conditions, etc). 
Once differentially expressed genes are identified, enrichment analysis (EA) methods can be conducted to identify groups of genes (e.g. particular pathways) that are differentially expressed, and offer insights into biological mechanisms. 
One example of such a method is the Gene Set Enrichment Analysis (GSEA), which is frequently used for high-throughput gene expression data analysis.

This course will cover GSEA and alternative enrichment tools. Because most of their implementations are directly linked to databases that annotate the function of genes in the cell, the course will also give an overview of functional annotation databases such as the Gene Ontology.

## Learning outcomes

At the end of the course, the participants will be able to:

1. Distinguish available enrichment analysis methods
2. Apply GSEA and over-representation analysis using R
3. Determine whether the genes of a GO term have a statistically significant difference in expression or not
4. Learn where to find other gene sets in databases (e.g. KEGG, oncogenic gene sets) and use them in R

## Exercises

The course is divided into small blocks, each with a lecture and relevant practical exercises. We will comment the practical exercises at the end of each block.

## Asking questions
During lectures, you are encouraged to ask questions. Just unmute yourself and ask your questions or use the Zoom functionality. Find the buttons in the participants list ('Participants' button):

<figure>
  <img src="assets/images/zoom_icons.png" width="300"/>
</figure>

Alternatively, (depending on your zoom version or OS) use the 'Reactions' button:

<figure>
  <img src="assets/images/reactions_zoom.png" width="200"/>
</figure>


